src/coreclr/vm/object.h

```diff
@@ -739,6 +739,78 @@ class PtrArray : public ArrayBase
 
 #define OFFSETOF__PtrArray__m_Array_              ARRAYBASE_SIZE
 
+/* Corresponds to the managed Span<T> and ReadOnlySpan<T> types.
+   This should only ever be passed from the managed to the unmanaged world byref,
+   as any copies of this struct made within the unmanaged world will not observe
+   potential GC relocations of the source data. As a result we've deleted ctors
+   which allow making byval copies of the span. */
+template < class KIND >
+class Span
+{
+private:
+    /* Keep fields below in sync with managed Span / ReadOnlySpan layout. */
+    KIND* _pointer;
+    unsigned int _length;
+
+public:
+    class ControlledMutabilityRef
```

Comment:
It is fine to modify the passed in Span - as long as you are following all other rules for writting manually managed code. I do not see why what it is trying to protect.

(In fact, I am worried that this may introduce a subtle hard-to-see GC holes by creating copy of the pointer that is not protected by the GC.)

Reply:
I was trying to pass `const Span<T>` (essentially, `ReadOnlySpan<T>`) into `InvokeArrayConstructor` and force a compilation error if somebody tried to reassign the ref. If this is too clever I can remove the code, and we simply annotate the type with "it's up the consumer to know if mutating this object is illegal, and we won't enforce any const-ness."

Reply:
I think it is trying to be too clever.

Reply:
(Also, I see this as throw-away code that will get deleted once we rewrite the reflection in C# except the lowest layer.)